sbi/neural_nets/embedding_nets.py

```diff
@@ -44,81 +45,151 @@ def forward(self, x: Tensor) -> Tensor:
-            First two layers are convolutional, followed by fully connected layers.
-            Performing 1d convolution and max pooling with preset configs.
+        """Convolutional embedding network.
+        First two layers are convolutional, followed by fully connected layers.
+
+        Automatically infers whether to apply 1D or 2D convolution depending on
+        input_shape.
+        Allows usage of multiple (color) channels by passing in_channels > 1.
 
         Args:
-            input_dim: Dimensionality of input.
-            output_dim: Dimensionality of the output.
-            num_conv: Number of convolutional layers.
-            num_fully_connected: Number fully connected layer, minimum of 2.
-            num_hiddens: Number of hidden dimensions in fully-connected layers.
-            out_channels_cnn_1: Number of oputput channels for the first convolutional
-                layer.
-            out_channels_cnn_2: Number of oputput channels for the second
-                convolutional layer.
+            input_shape: Dimensionality of input, e.g., (28,) for 1D, (28, 28) for 2D.
+            in_channels: Number of image channels, default 1.
+            out_channels_per_layer: Number of out convolutional out_channels for each
+                layer. Must match the number of layers passed below.
+            num_cnn_layers: Number of convolutional layers.
+            num_linear_layers: Number fully connected layer.
+            num_linear_units: Number of hidden units in fully-connected layers.
+            output_dim: Number of output units of the final layer.
             kernel_size: Kernel size for both convolutional layers.
             pool_size: pool size for MaxPool1d operation after the convolutional
                 layers.
-
-            Remark: The implementation of the convolutional layers was not tested
-            rigourously. While it works for the default configuration parameters it
-            might cause shape conflicts fot badly chosen parameters.
         """
-        super().__init__()
-        self.input_dim = input_dim
-        self.output_dim = output_dim
-        self.num_hiddens = num_hiddens
-
-        # construct convolutional-pooling subnet
-        pool = nn.MaxPool1d(pool_size)
-        conv_layers = [
-            nn.Conv1d(1, out_channels_cnn_1, kernel_size, padding="same"),
-            nn.ReLU(),
-            pool,
-            nn.Conv1d(
-                out_channels_cnn_1, out_channels_cnn_2, kernel_size, padding="same"
-            ),
-            nn.ReLU(),
-            pool,
-        ]
-        self.conv_subnet = nn.Sequential(*conv_layers)
+        super(CNNEmbedding2D, self).__init__()
 
-        # construct fully connected layers
-        input_dim_fc = out_channels_cnn_2 * (int(input_dim / out_channels_cnn_2))
+        assert isinstance(
+            input_shape, Tuple
+        ), "input_shape must be a Tuple of size 1 or 2, e.g., (width, [height])."
```

Comment:
why are there square brackets around height?

Reply:
to show that one perform 1D convolution too, i.e., passing height is optional.